benchopt/plotting/generate_html.py

```diff
@@ -0,0 +1,334 @@
+        for each dataset.
+    """
+    dataset_names = df['data_name'].unique()
+    objective_names = df['objective_name'].unique()
+    fname_short = fname.replace('outputs/', '').replace('/', '_')
+
+    figures = {}
+    n_figure = 0
+    for data_name in dataset_names:
+        figures[data_name] = {}
+        df_data = df[df['data_name'] == data_name]
+        for objective_name in objective_names:
+            figures[data_name][objective_name] = {}
+            df_obj = df_data[df_data['objective_name'] == objective_name]
+
+            for k in kinds:
+                if k not in PLOT_KINDS:
+                    raise ValueError(
+                        f"Requesting invalid plot '{k}'. Should be in:\n"
+                        f"{PLOT_KINDS}")
+                plot_func = globals()[PLOT_KINDS[k]]
+                try:
+                    fig = plot_func(df_obj, plotly=True)
+                except TypeError:
+                    fig = plot_func(df_obj)
+                figures[data_name][objective_name][k] = export_figure(
+                    fig, f"{fname_short}_{n_figure}"
+                )
+                n_figure += 1
+
+    return dict(
+        figures=figures, dataset_names=dataset_names, fname_short=fname_short,
+        objective_names=objective_names, kinds=list(kinds)
+    )
+
+
+def export_figure(fig, fig_name):
+    """
+    Export figure to svg
+
+    Parameters
+    ----------
+    fig : plotly.graph_objs._figure.Figure
+        Figure from plotly
+
+    fig_name : str
+        Name to be given to the figure.
+
+    Returns
+    -------
+    str
+        Path to the figure.
+    """
+    print(type(fig))
+    if hasattr(fig, 'to_html'):
+        return fig.to_html(include_plotlyjs=False)
+
+    fig_basename = f"{fig_name}.svg"
+    save_name = HTML_DIR_FIGURES / fig_basename
+    fig.savefig(save_name)
+    plt.close(fig)
+    return f'figures/{fig_basename}'
+
+
+def get_result_from_df(df, kinds):
+    """
+    Generate figure and CSV file from dataframe
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+        pandas DataFrame containing the benchmark results
+    kinds : list of str
+        List of the kind of plots that will be generated. This needs to be a
+        sub-list of PLOT_KINDS.keys().
+
+    Returns
+    -------
+    dict
+        Dictionary containing to results
+    """
+    results = []
+    HTML_DIR_OUTPUTS.mkdir(exist_ok=True, parents=True)
+    HTML_DIR_FIGURES.mkdir(exist_ok=True, parents=True)
+
+    datasets = list(df['data_name'].unique())
+    fname = "benchmark_benchopt_run_" + str(datetime.now())[:19] + ".csv"
+    fname = fname.replace(" ", "_").replace(":", "_")
+
+    # Generate figures
+    result = dict(
+        fname=fname,
+        datasets=datasets,
+        **generate_plot_benchmark(df, kinds, fname)
+    )
+    results.append(result)
+
+    for result in results:
+        result['page'] = \
+            f"{str(result['fname_short']).replace('.csv', '.html')}"
```

Comment:
This seems weird as `results` is guaranteed to have only 1 element no?
Maybe remove the `results` and return `[result]`?

Reply:
Also, most of the code can be factorized with `get_results`.
It would maybe be better to have an option in `get_results` that allow to pass directly a dataframe and compute the associated element list of fnames?

Reply:
Actually, it is not the same function (it does not take a list of dataframes but a list of csv files, the comments are not accurate). I already tried to factorize and I remember it was tricky.